[CS] A couple of fixes for implicit callAsFunction

hamishknight opened the pull request:
Start resolving callees for an implicit call to `callAsFunction` by adding a case for it in `getCalleeLocator`, which allows it to be used with default arguments and diagnosed with things like requirement failures.

In addition, prevent callables from being used with `AnyObject` lookup, previously this hit an assert in CSApply, now we'll reject it with an error.

Finally, tweak the CSApply logic for `callAsFunction` such that we apply all of `finishApply`'s transforms to it, which handles things like IUO unwraps.

Resolves SR-11386, SR-11778, SR-11829, and SR-11881.

hamishknight committed:
[Sema] Add TypeBase::isCallableNominalType

This checks whether a type supports being called
via callAsFunction. A request is used in order to
cache the result of this query.

hamishknight committed:
[CS] Resolve callees for implicit callAsFunction

Add a case to getCalleeLocator to return the
appropriate locator for a call to an implicit
callAsFunction member reference.

Resolves SR-11386 & SR-11778.

hamishknight committed:
[CS] Apply all finishApply's transforms to callAsFunction

Rather than coercing the call arguments ourselves,
just build the finished member reference for the
callAsFunction method, and let finishApply do the
rest. This allows us to deal with transformations
such as IUO unwraps.

Resolves SR-11881.

hamishknight commented:
@swift-ci please test

xedin commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -497,6 +509,25 @@ ConstraintSystem::getCalleeLocator(ConstraintLocator *locator,
   return getConstraintLocator(anchor);
 }
 
+bool ConstraintSystem::isCallableNominalType(Type ty) {
```
This kind of feels like it it belongs on the type (or even a nominal declaration) itself more than a part of constraint system interface...

hamishknight replied:
@xedin My original thought was to put it on the constraint system to benefit from its cache of member lookups, but I guess we just make a request for it instead. I've gone ahead and updated things so that `isCallableNominalType` is now a member of `TypeBase`, does that look good?

Thanks for the review!

xedin replied:
Looks great, thank you!

xedin replied:
Looks great, thank you!

xedin reviewed (approved):
Thank you!

hamishknight committed:
Don't check access control for isCallableNominalType

We still want the constraint system to try looking up
a callAsFunction method even if it's inaccessible, as
we'll be able to record a fix and emit a suitable
diagnostic.

hamishknight committed:
[CS] Prevent callAsFunction from being used on AnyObject

Previously this hit an assertion in CSApply.
While we could technically make it work, it's
probably best not to allow it.

Resolves SR-11829.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
Rebased to resolve a merge conflict

@swift-ci please smoke test and merge

hamishknight commented:
@swift-ci please smoke test and merge

hamishknight commented:
@swift-ci please smoke test and merge
